### PR TITLE
414 virtual-scroll dynamic height

### LINKED ISSUE #414: VirtualScroll: Dynamic Height
**Describe the bug**
The way the virtual scroll calculates height isn't very dynamic, It would be good that if the height of the list view changes (view css height 100 and resizing the page) then the virtual scroll height is auto calculated.

**To Reproduce**

1. Go to http://localhost:4300/ids-list-view/
2. Change the height of the card/listview so it is higher (or more dynamic)
3.  See error

### PR DESCRIPTION
**Pull request by vivz753**
**Explain the details for making this change. What existing problem does the pull request solve?**

Please see comments on #414 

**Related github/jira issue (required)**:

- Closes #414 

**Steps necessary to review your pull request (required)**:

### **CHANGES with this PR:**
http://localhost:4300/ids-virtual-scroll
Screen Shot 2021-11-08 at 5 26 20 PM
Screen Shot 2021-11-08 at 5 26 28 PM

http://localhost:4300/ids-data-grid
Screen Shot 2021-11-08 at 5 15 23 PM
Screen Shot 2021-11-08 at 5 15 30 PM

http://localhost:4300/ids-list-view
Screen Shot 2021-11-08 at 5 15 49 PM
Screen Shot 2021-11-08 at 5 15 57 PM

***

### **PREVIOUSLY, _no responsive height at all_:**
Screen Shot 2021-11-08 at 5 20 06 PM
Screen Shot 2021-11-08 at 5 20 16 PM
Screen Shot 2021-11-08 at 5 20 28 PM

**Included in this Pull Request**:
- [ ] An e2e or functional test for the bug or feature.
- [ ] A note to the change log.

### CODE REVIEW

File: src/components/ids-virtual-scroll/ids-virtual-scroll.js [hunk 1]
```diff
@@ -110,16 +108,21 @@ class IdsVirtualScroll extends mix(IdsElement).with(IdsEventsMixin) {
    * @private
    */
   applyHeight() {
-    const viewport = this.container.querySelector('.ids-virtual-scroll-viewport');
+    const content = this.container.querySelector('.ids-virtual-scroll-content');
```
Comment (vivz753) [inline src/components/ids-virtual-scroll/ids-virtual-scroll.js:111]: Viewport is actually the outermost div, `this.container`

Renamed div accordingly.

**Review by tmcconechy — CHANGES_REQUESTED** [review]
This looks really solid. Just a few things mainly around examples/not doing this all the time.

1. the ability to resize down should be an option in datagrid and listview (maybe fillToBottom ?or fill-height similar to the card or something?)
2. the size should go to the bottom - the grid gutter size (16px but configurable)
3. lets show this on dedicated example pages for listview and datagrid (can maybe leave the virtual scroll example) i.e. data-grid/fill-to-bottom.html or something. Need to think about the name? 
4. the card header shouldnt be hidden as it is needed.

**Comment by vivz753** [comment]
> This looks really solid. Just a few things mainly around examples/not doing this all the time.
> 
> 1. the ability to resize down should be an option in datagrid and listview (maybe fillToBottom ?or fill-height similar to the card or something?)
> 2. the size should go to the bottom - the grid gutter size (16px but configurable)
> 3. lets show this on dedicated example pages for listview and datagrid (can maybe leave the virtual scroll example) i.e. data-grid/fill-to-bottom.html or something. Need to think about the name?
> 4. the card header shouldnt be hidden as it is needed.

1. By default, it will use 100vh and respond to screen size. Are you saying you want a fixed px height to be the default? Or you just want an option to disable the responsiveness? The responsiveness is easily overridden when user specifies height, i.e. `height="300px"`. I don't really see the need to add the attribute for the responsive height but can understand if you want to make the fixed height default, and have an `auto-height` attribute override it.
2. Wdym the size should go to the bottom? As in the container should be flush with the bottom of the browser?
3. How about Responsive Data Grid/List View for the dedicated example? 
4. The card has css for the `auto-height` attribute which deliberately hides the header. From 5 months ago:  https://github.com/infor-design/enterprise-wc/blob/15a39a97f11238b2cba640bd346666251844b7b9/src/ids-card/ids-card.scss#L20

File: demos/ids-data-grid/example.html [hunk 2]
```diff
@@ -3,7 +3,7 @@
 </ids-layout-grid>
 <ids-layout-grid auto="true">
   <ids-layout-grid-cell>
-    <ids-data-grid id="data-grid-1" virtual-scroll="false" label="Books" row-height="medium">
+    <ids-data-grid id="data-grid-1" virtual-scroll="true" label="Books" row-height="medium">
```
[the remark this answers is not in the document]
Comment (vivz753) [reply demos/ids-data-grid/example.html:6]: Sure, I can change it back

**Comment by tmcconechy** [comment]
Oh 👍🏻 

1. As per requirements before card and datagrid should be have a default static height (the card height is static) and the default for datagrid is the number of rows. Then add an option to make it "expand down" to the bottom (i.e. responsive) thats the non default i.e.
https://main-enterprise.demo.design.infor.com/components/datagrid/example-index
vs
https://main-enterprise.demo.design.infor.com/components/datagrid/example-fixed-header.html

2. When it "expands down" it should go all the way to the bottom (16px on left/right bottom) like this example https://main-enterprise.demo.design.infor.com/components/datagrid/example-fixed-header.html (the grid was 20px now should be 16px) but same idea
3. Sure sounds good. To consider (and im torn on the name here it was "full-height") but maybe (responsive = "true" is good?). but its just the height so thats whats throwing me off here...
4. Oh i see. Yes that makes a card that has no header and is the size of the contents. I was thinking we need one that has a header and expands down to the bottom (so maybe tied to this new option in 3 instead?)

**Comment by vivz753** [comment]
1. I can easily revert the list-view/data-grid examples back to their fixed heights, so I have done that. Are you requesting a feature for data-grid to be able to expand down upon toggling some attribute like `auto-height`? That is possible, and can easily be done. Let's separate that onto another ticket. ... On second thought, it doesn't make sense to add an attribute for responsive height when that style only exists tentatively if `virtual-scroll="true"`. In that case I would suggest spec'ing this out a bit more... in the absence of virtual-scroll, it would make sense to add styling for responsive height in the case that ids-data-grid is using regular divs too. Do you want me to make a ticket for this?
2. I have already made it so that data grid goes to the bottom, and it looks just like your example. It might not look like it because there are only 9 rows so the rest is empty space. Can you look again to make sure we're on the same page? [screenshot](https://user-images.githubusercontent.com/22609861/140843890-3cf1b194-13de-4af8-9c2e-24b92e45dc79.png)
3. I thought we were just talking about the name for the example html file. Are you suggesting a different name for the attribute? I would go with `auto-height` for consistency since ids-card uses that.
4. What would this look like in terms of changing the existing `auto-height` attribute in ids-card? Would this be a modification to the auto-height css or are you suggesting a new attribute that acts like auto-height and also includes the header? Sounds like another ticket for a feature request for ids-card.

**Comment by tmcconechy** [comment]
lets try this one more and see. i think im not explaining myself too well. So if any more questions lets connect on MS Teams and chat.

1. I think this is not what im thinking. The virtual-scroll option should be independent of the height. But i think a height option should work for the case. So no height its what it is now in datagrid and listbiew). `height="100vh"` or something should expand down to fill the parent (minus padding). So basically just because virtual-scroll is on doesnt mean it should expand down without a second option. Does that need a third issue or does that just confuse things?
2. Yes that screen shot looks correct. last time i saw it it was maybe 50+ pixels from the bottom. But also it should only do this if an option is set (like `height` now)
3. Yes, what about height="100%" or height="400" or height="100vh" the thing with `auto-height` is what it is doing is height="auto" so it means make it the size of the contents. So not the same thing. But i dont mind refactoring that name to maybe `height="auto"` ?
4. Since auto-height makes the card the exact size of the contents. i think what i said on `height="auto"` makes sense. Maybe it should work independently of if there is a header or not. So basically will be the height of contents and you have to put the header in the page or not. it will look exactly the same as you have it but showing a header....

Im sure thats more confusing.. But i think we are getting there.

**Comment by vivz753** [comment]
I think I understand what you're saying, but just to clarify...

Requirements for ids-data-grid:
- when virtual-scroll is enabled, height should stay the same as data-grid's default fixed height i.e. 300 px
- virtual-scroll should NOT have responsive height/expand to the bottom by default. It must be specified with `height="100vh"` or a dedicated attribute i.e. `full-height`

I was trying to say that ids-data-grid currently doesn't have any code for responsive height styling for the regular divs (when virtual-scroll="false") so that needs to get added in. That's out of scope for this ticket because this issue is regarding dynamic height for ids-virtual-scroll, not ids-data-grid. So if you want me to do that, it would have to be on another ticket.

Can you please be more specific with the requirements for ids-card? I think what you're saying is that you want `auto-height` to check first is there is a header and then account for the height of it and use the remaining space to fill it with contents, so that it shows both the header AND the contents AND has a responsive height. Currently it _does_ work independently whether there is a card header or not, it just handles it by hiding the header completely. It sounds like you want to change this design. Just to be clear, this is also out of scope for this PR, so sounds like another ticket needs to be made.

Please confirm, and I will go ahead and make the tickets.

Let's merge this one in first and I'll get started on the others.

**Comment by tmcconechy** [comment]
Will make a new issue for this...

Virtual Scroll -> Dynamic by default but can pick up fixed height
Card -> 360px  by default - can set to auto-fit or auto-height
List View -> auto-fit by default and nothing else right now (card or page of the parent)
Datagrid -> Number of rows or auto-fit (future option)

Future Ideas:
```
height="auto" -> auto-fit
height="100vh" -> fit down? but from screen
height="100%" ->  fit down? but from parent
height="400px" -> fixed??
```

Other rejected name ideas?
```
autogrow = auto-height
full-height
content-fit
size-to-parent
fit-parent
fit-content
size-parent
size-content?
```

**Comment by vivz753** [comment]
^ Summary of topics discussed over phone call with me and Tim.

Changes made to match heights of the ids-virtual-scroll examples by dynamically calculating space from top to subtract from 100vh along with 16px margin

Updated screenshot of ids-virtual-scroll examples:
Screen Shot 2021-11-10 at 2 26 47 PM

**Review by jmacaluso711 — CHANGES_REQUESTED** [review]
@vivz753 I'm seeing a few issues when setting a fixed height on the virtual scroll component

Double scroll bar when the parent `ids-card` is not set to `auto-height` or has a fixed height set:
Screen Shot 2021-11-11 at 3 46 30 PM

Missing card header with set to `auto-height`:
Screen Shot 2021-11-11 at 3 48 39 PM

**Comment by vivz753** [comment]
^ Regarding comments, @jmacaluso711 you are entirely correct. 

Whenever you set a custom height for virtual-scroll, you MUST set `auto-height` for ids-card. ids-card by default has a fixed height of 360px, 50px for the header, 310px for the contents, so naturally putting anything in there that is not exactly equal to 310px will cause a scroll bar to appear hence the double scrollbars. `auto-height` however allows the content container of the card to grow dynamically according to what's inside of it.

`auto-height` when toggled deliberately adds a CSS class that sets the header `display: none`. Whether or not this is an intended feature for design needs to be discussed, and is out of scope for this issue however.

**Comment by tmcconechy** [comment]
If i mentioned this already forgive me as there is a lot of chat in this one. But from what i think is a good jump off point.

1) We make http://localhost:4300/ids-virtual-scroll the same as it was before so visually its unchanged? Right now the header is missing and it expands down further than the bottom of the page.
Screen Shot 2021-11-12 at 10 33 57 AM
2) We make two new examples to show it expanding down in ids-list-view/dynamic-height and ids-data-grid/dynamic-height because i think the virtual scroll example looks a bit wierd now.

**Comment by vivz753** [comment]
@tmcconechy the examples on ids-virtual-scroll are the same height, and they fit the page, you just need to scroll down a bit.

Screen Shot 2021-11-15 at 9 26 06 AM

I can still make seperate example pages if you want. Let me know and I'll make sure to add it in.

**Comment by tmcconechy** [comment]
@vivz753 OK, then i guess we should just adjust it so I dont have to  scroll down a bit. Can we just make it fill to the bottom -16px of the page no matter the page size? Seems like it needs to take the position from the top into account.

Screen Shot 2021-11-15 at 12 53 02 PM

**Comment by vivz753** [comment]
The scrolling down is a result of header being there and taking up space, which IIRC might be removed in the future, so I don't think it makes sense to write code to account for that. It technically does fit to the bottom of the page -16px, and resizes based on 100vh.

**Comment by tmcconechy** [comment]
Yeah but the point is it doesnt look right, so i guess from that standpoint it does make sense to fix to me.

So i guess either we add code to adjust for how far it is down the page or put it in a parent container sized that looks right? Or leave it as it was before with a static height?

Really the last remaining thing for me...

**Comment by tmcconechy** [comment]
I came up with another idea which is to take... demos/ids-virtual-scroll/index.html and remove the theme switcher but the height still isnt quite right. Maybe for now we can do this and adjust the height for the example accordingly for now.

```
{{> ../layouts/head-themeless.html }}
{{> example.html }}
{{> ../layouts/footer.html }}
```

**Comment by vivz753** [comment]
Updated examples

Screen Shot 2021-11-16 at 8 50 46 AM

**Review by tmcconechy — APPROVED** [review]

**Review by jmacaluso711 — APPROVED** [review]

**Pull request merged**